src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -1374,7 +1377,9 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         asm.Type.getMethodDescriptor(asmType(functionalInterface), argsArray:_*)
       }
       val lambdaParams = lambdaTarget.paramss.head.drop(numCaptured)
-      val constrainedType = MethodBType(lambdaParams.map(p => typeToBType(p.tpe)), typeToBType(lambdaTarget.tpe.resultType)).toASMType
+      val lambdaParamsBTypes = new Array[BType](lambdaParams.size)
+      lambdaParams.iterator.map(p => typeToBType(p.tpe)).copyToArray(lambdaParamsBTypes)
```

Comment:
You could compute the size as a side-effect of `map` if saving an iteration is relevant. Maybe add some helper methods `mapToArray`, `toArray(size: Int)`?

Reply:
I've added some helpers. It is necessary to iterate twice to create an array of the right size to hold the result elements.